addons/nuxeo-retention/nuxeo-retention-core/src/test/java/org/nuxeo/retention/test/RetentionTestCase.java

```diff
@@ -0,0 +1,155 @@
+@Deploy("org.nuxeo.retention.core:OSGI-INF/retention-listeners.xml")
+@Deploy("org.nuxeo.retention.core:OSGI-INF/retention-operations.xml")
+public class RetentionTestCase {
+
+    @Inject
+    protected RetentionManager service;
+
+    @Inject
+    protected CoreFeature coreFeature;
+
+    @Inject
+    protected CoreSession session;
+
+    @Inject
+    protected BulkService bulkService;
+
+    protected DocumentModel file;
+
+    @Before
+    public void setup() {
+        file = session.createDocumentModel("/", "File", "File");
+        file = session.createDocument(file);
+        file = session.saveDocument(file);
+    }
+
+    protected void assertStillUnderRetentionAfter(DocumentModel doc, RetentionRule rule, int timeoutMillis)
+            throws InterruptedException {
+        doc = service.attachRule(doc, rule, session);
+        assertTrue(doc.isRecord());
+
+        awaitRetentionExpiration(1_000);
+
+        doc = session.getDocument(doc.getRef());
+
+        // it is still under retention and has a retention date
+        assertTrue(session.isUnderRetentionOrLegalHold(doc.getRef()));
+        assertNotNull(session.getRetainUntil(doc.getRef()));
+    }
+
+    protected void awaitRetentionExpiration(long millis) throws InterruptedException {
+        // wait a bit more than retention period to pass retention expiration date
+        coreFeature.waitForAsyncCompletion();
+        Thread.sleep(millis);
+        // trigger manually instead of waiting for scheduler
+        new RetentionExpiredFinderListener().handleEvent(null);
+        assertTrue("Bulk action didn't finish", bulkService.await(Duration.ofSeconds(60)));
+        coreFeature.waitForAsyncCompletion();
+    }
+
+    protected RetentionRule createRuleWithActions(RetentionRule.ApplicationPolicy policy,
+            StartingPointPolicy startingPointPolicy, List<String> docTypes, String startingPointEventId,
+            String startingPointExpression, String matadataXPath, long years, long months, long days,
+            long durationMillis, List<String> beginActions, List<String> endActions) {
+        DocumentModel doc = session.createDocumentModel("/RetentionRules", "testRule", "RetentionRule");
+        RetentionRule rule = doc.getAdapter(RetentionRule.class);
+        rule.setDurationYears(years);
+        rule.setDurationMonths(months);
+        rule.setDurationDays(days);
+        rule.setApplicationPolicy(policy);
+        rule.setStartingPointPolicy(startingPointPolicy);
+        rule.setDocTypes(docTypes);
+        rule.setStartingPointEvent(startingPointEventId);
+        rule.setStartingPointExpression(startingPointExpression);
+        rule.setMetadataXpath(matadataXPath);
+        rule.setDurationMillis(durationMillis);
+        rule.setBeginActions(beginActions);
+        rule.setEndActions(endActions);
+        doc = session.createDocument(doc);
+        return session.saveDocument(rule.getDocument()).getAdapter(RetentionRule.class);
```

Comment:
i am wondering if we replace the whole method by a `Builder` to avoid passing `0` values or an empty list ... 
```
    protected class RetentionRuleBuilder {

        protected DocumentModel documentModel;

        protected RetentionRule rule;

        protected RetentionRuleBuilder() {
            DocumentModel doc = session.createDocumentModel("/RetentionRules", "testRule", "RetentionRule");
            rule = doc.getAdapter(RetentionRule.class);
        }

        RetentionRuleBuilder durationYears(long years) {
            rule.setDurationYears(years);
            return this;
        }

    }
....
```

Reply:
It's ok for tests